### PR TITLE
Feature: Add model-joints optional parameter

### PR DESCRIPTION
**Pull request by Yeshasvitvs**
In the light of https://github.com/robotology/idyntree/issues/527#issuecomment-491531318 this PR adds a small patch to yarprobotstatepublisher by adding a `model-joints` parameter (may be not an ideal name) which is set to `greater` by default to ensure back compatibility. This patch is tested with another robot model when the number of joint models is lower than the joints streamed in the joint_states topic. 

@lrapetti @traversaro

### CODE REVIEW

**Comment by traversaro** [comment]
As mentioned in https://github.com/robotology/idyntree/issues/527#issuecomment-491543874 , I think it would be better to first address the reviews by @prashanthr05  on https://github.com/robotology/idyntree/pull/513 (and getting it merged to `devel`) before adding new commits to the same PR. If you just need a fork of iDynTree for internal temporary use, I suggest that you fork the repo in some location in which you have access (a personal repo or the lab's organisation) and push the branch here.

**Comment by traversaro** [comment]
I still strongly believe that you should just finish https://github.com/robotology/idyntree/pull/513 and track the latest iDynTree `devel` branch, but if you really need to fork iDynTree, you now have the repo https://github.com/robotology/idyntree/pull/513 on which @Yeshasvitvs and @lrapetti have full access.

**Comment by traversaro** [comment]
Note: I edited the previous comment to fix the link to https://github.com/dic-iit/idyntree-hde-fork .

**Comment by traversaro** [comment]
What is the status of this PR? @Yeshasvitvs @lrapetti

**Comment by Yeshasvitvs** [comment]
@traversaro this PR is important to run rviz visualization with the real robot. I changed the base to devel and I see some conflicts. I will clean it up.

File: src/tools/yarprobotstatepublisher/src/main.cpp [hunk 1]
```diff
@@ -32,14 +32,15 @@ int main(int argc, char *argv[])
     // print available command-line options
     // upon specifying --help; refer to xml
     // descriptor for the relative documentation
-    if (rf.check("help"))
+    iif (rf.check("help"))
```
Comment (prashanthr05) [inline src/tools/yarprobotstatepublisher/src/main.cpp]: iif -> if

**Comment by prashanthr05** [comment]
Tested this `model-joints` feature with option `greater`, and was working fine for me.
Thanks @Yeshasvitvs

**Comment by Yeshasvitvs** [comment]
@traversaro This PR is important to use `yarprobotstatepublisher` with real rviz robot model. It updates the size check between the model passed and the size of the joints streamed in `/joint_states` topic.

**Comment by traversaro** [comment]
The merge of https://github.com/robotology/idyntree/pull/548 created problems in this PR. Feel free to fix them and ping me when the PR is ready for review, thanks!

**Comment by Yeshasvitvs** [comment]
@traversaro fixed the conflicts after the merge of https://github.com/robotology/idyntree/pull/548
You can proceed to review.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp [hunk 2]
```diff
@@ -233,17 +251,24 @@ void YARPRobotStatePublisherModule::onRead(yarp::rosmsg::sensor_msgs::JointState
...
+        if (jntIndex == iDynTree::JOINT_INVALID_INDEX)
             continue;
 
         m_jointPos(model.getJoint(jntIndex)->getDOFsOffset()) = v.position[i];
     }
 
     // Set the updated joint positions
     m_kinDynComp.setJointPos(m_jointPos);
-
+
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]:
```suggestion

```

File: src/tools/yarprobotstatepublisher/include/robotstatepublisher.h [hunk 3]
```diff
@@ -23,8 +23,10 @@
 #include <iDynTree/KinDynComputations.h>
 
 #include <yarp/rosmsg/sensor_msgs/JointState.h>
+
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/include/robotstatepublisher.h]:
```suggestion
```

File: src/tools/yarprobotstatepublisher/include/robotstatepublisher.h [hunk 4]
```diff
@@ -23,8 +23,10 @@
...
 #include <memory>
 
+
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/include/robotstatepublisher.h]:
```suggestion
```

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp [hunk 5]
```diff
@@ -255,16 +280,51 @@ void YARPRobotStatePublisherModule::onRead(yarp::rosmsg::sensor_msgs::JointState
...
-        if(m_baseFrameIndex == frameIdx)    // skip self-tranform
-            continue;
-
-        iDynTree::Transform base_H_frame = m_kinDynComp.getRelativeTransform(m_baseFrameIndex, frameIdx);
-        iDynTree::toYarp(base_H_frame.asHomogeneousTransform(), m_buf4x4);
-        m_iframetrans->setTransform(m_tfPrefix + model.getFrameName(frameIdx),
-                                    m_tfPrefix + model.getFrameName(m_baseFrameIndex),
-                                    m_buf4x4);
+    // Publish the frames on TF
+    bool m_publishGlobalTF = false; // TODO
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: Just to understand, why does this part of the code change? Is this related to the `model-joints` parameter?

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp [hunk 6]
```diff
@@ -233,17 +251,24 @@ void YARPRobotStatePublisherModule::onRead(yarp::rosmsg::sensor_msgs::JointState
...
-
-        if (!(model.getJoint(jntIndex)->getNrOfDOFs()) || jntIndex == iDynTree::JOINT_INVALID_INDEX)
+        for (size_t j=0; j < model.getNrOfDOFs(); j++)
+        {
+            if (v.name[i] == model.getJointName(j))
+            {
+                jntIndex = model.getJointIndex(v.name[i]);
+                break;
+            }
+        }
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: What is the purpose of this change?

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp [hunk 7]
```diff
@@ -217,12 +221,26 @@ void YARPRobotStatePublisherModule::onRead(yarp::rosmsg::sensor_msgs::JointState
...
-                 << " joints, while the received JointState message has " << v.name.size() << " joints.";
-        return;
+    if (m_model_joints == "lower") {
+
+        // Check if the joints states are more than the number of joints of the model
+        if (v.name.size() > m_jointPos.size()) {
+            yError() << "Size mismatch. Model has " << m_jointPos.size()
+                     << " joints, while the received JointState message has " << v.name.size() << " joints.";
+            return;
+        }
```
Comment (traversaro) [inline src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: If this is the case, in this case, what is the value contained in `m_jointPos` for joints that are not contained in the joint state topic? To avoid confusion, I think we should document it.

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 6 above)
Comment (Yeshasvitvs) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: My bad, I didn't resolve the conflicts correctly after merging #548 to devel!  
Updated in https://github.com/robotology/idyntree/pull/533/commits/fd6f66f37697ba5ef6d524b825afb076c6cfd4e5

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 5 above)
Comment (Yeshasvitvs) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: My bad, I didn't resolve the conflicts correctly after merging #548 to devel!  
Updated in https://github.com/robotology/idyntree/pull/533/commits/fd6f66f37697ba5ef6d524b825afb076c6cfd4e5

File: src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp (hunk 7 above)
Comment (Yeshasvitvs) [reply src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp]: @traversaro at the moment I am not certain how and when this case occurred but @lrapetti and I ran into this problem a while back with another robot model we were using for retargeting. Unfortunately, we cannot recall the exact steps to reproduce the problem.

**Comment by Yeshasvitvs** [comment]
> Tested this `model-joints` feature with default option `greater`, and was working fine for me.
> Thanks @Yeshasvitvs
> The segmentation fault was also fixed, with this update.

The segmentation fault happens when trying to get the robot joint index based on the joint name present in the `/joint_states` topic.
https://github.com/robotology/idyntree/blob/2cbb75f67a8a9b93428d93d941c8d1b65ba6a415/src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp#L236

On the real robot (`iCubGenova04`), the `/joint_states` topic consists of **50** joints, including the joint values of the fingers, where as the `model.urdf` of `iCubGenova04` doesn;t have the hands. So, when trying to retrieve a finger joint based from the mode, segmentation fault occurs.

@prashanthr05 @lrapetti @traversaro

**Comment by Yeshasvitvs** [comment]
Given that this PR has been opened a while ago and a lot of new changes are merged to devel #538, for the sake of a clean history this PR is not longer relevant. I prefer to open a new PR that addresses the problem highlighted in https://github.com/robotology/idyntree/pull/533#issuecomment-518314597 and also remove the joint size check https://github.com/robotology/idyntree/blob/devel/src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp#L220 as discussed with @traversaro
